### ISSUE TITLE
Failure with NCCL 2.10 + CUDA 11.4

### ISSUE THREAD

**Issue by xkszltl** [NVIDIA/nccl-tests#91]
We're seeing strange failure when updating to CUDA 11.4 with our submodule of nccl-tests.

```
HEAD is now at 44df0bf Merge pull request #88 from nzmsv/master                                           
root@66cf90142aaf:/nccl-tests# LD_LIBRARY_PATH="/usr/local/openmpi/lib:$LD_LIBRARY_PATH" make -j$(nproc) MPI=0 MPI_HOME="/usr/local/openmpi" NVCC_GENCODE="$(echo 52 60 61 70 75 80 86 | xargs -rn1 | xargs -rI{} echo -gencode=arch=compute_{},code=sm_{} | p
aste -sd' ' -) -gencode=arch=compute_80,code=compute_80"                                                  
make -C src build                                                                                         
make[1]: Entering directory '/nccl-tests/src'                                                             
Compiling  all_reduce.cu                       > ../build/all_reduce.o                                    
Compiling  common.cu                           > ../build/common.o                                        
Compiling  all_gather.cu                       > ../build/all_gather.o                                    
Compiling  broadcast.cu                        > ../build/broadcast.o                                     
Compiling  reduce_scatter.cu                   > ../build/reduce_scatter.o                                
Compiling  reduce.cu                           > ../build/reduce.o
Compiling  alltoall.cu                         > ../build/alltoall.o
Linking  ../build/all_reduce.o               > ../build/all_reduce_perf
Linking  ../build/all_gather.o               > ../build/all_gather_perf
Linking  ../build/broadcast.o                > ../build/broadcast_perf
Linking  ../build/reduce_scatter.o           > ../build/reduce_scatter_perf
Linking  ../build/reduce.o                   > ../build/reduce_perf
Linking  ../build/alltoall.o                 > ../build/alltoall_perf
make[1]: Leaving directory '/nccl-tests/src'
root@66cf90142aaf:/nccl-tests# LD_LIBRARY_PATH="/usr/local/openmpi/lib:$LD_LIBRARY_PATH" NCCL_DEBUG=INFO NCCL_NSOCKS_PERTHREAD=8 NCCL_SOCKET_NTHREADS=4 nvidia-smi -L | wc -l | xargs seq | xargs -n1 build/all_reduce_perf -b8 -e1G -f2 -g
# nThread 1 nGpus 1 minBytes 8 maxBytes 1073741824 step: 2(factor) warmup iters: 5 iters: 20 validation: 1
#
# Using devices
#   Rank  0 Pid   5869 on 66cf90142aaf device  0 [0x00] Tesla V100-SXM2-32GB
#
#                                                     out-of-place                       in-place
#       size         count    type   redop     time   algbw   busbw  error     time   algbw   busbw  error
#        (B)    (elements)                     (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)
           8             2   float     sum     3.38    0.00    0.00  1e+00     0.47    0.02    0.00  1e+13
          16             4   float     sum     3.33    0.00    0.00  1e+00     0.47    0.03    0.00  1e+13
          32             8   float     sum     3.23    0.01    0.00  1e+13     0.47    0.07    0.00  1e+00
          64            16   float     sum     3.31    0.02    0.00  7e+31     0.47    0.14    0.00  7e+31
         128            32   float     sum     3.31    0.04    0.00  7e+31     0.47    0.28    0.00  7e+31
         256            64   float     sum     3.32    0.08    0.00  7e+31     0.47    0.55    0.00  7e+31
         512           128   float     sum     3.32    0.15    0.00  3e+37     0.47    1.10    0.00  3e+37
        1024           256   float     sum     3.28    0.31    0.00  3e+37     0.46    2.23    0.00  3e+37
        2048           512   float     sum     3.22    0.64    0.00  3e+38     0.46    4.45    0.00  3e+38
        4096          1024   float     sum     3.25    1.26    0.00  3e+38     0.47    8.71    0.00  3e+38
        8192          2048   float     sum     3.32    2.47    0.00  3e+38     0.46   17.81    0.00  3e+38
       16384          4096   float     sum     3.19    5.14    0.00  3e+38     0.47   35.23    0.00  3e+38
       32768          8192   float     sum     3.32    9.88    0.00  3e+38     0.47   69.72    0.00  3e+38
       65536         16384   float     sum     3.38   19.42    0.00  3e+38     0.47  140.94    0.00  3e+38
      131072         32768   float     sum     3.31   39.66    0.00  3e+38     0.47  278.85    0.00  3e+38
      262144         65536   float     sum     3.44   76.20    0.00  3e+38     0.47  563.75    0.00  3e+38
      524288        131072   float     sum     4.02  130.42    0.00  3e+38     0.47  1115.39    0.00  3e+38
     1048576        262144   float     sum     6.00  174.91    0.00  3e+38     0.47  2231.01    0.00  3e+38
     2097152        524288   float     sum     8.66  242.16    0.00  3e+38     0.47  4510.00    0.00  3e+38
     4194304       1048576   float     sum    14.48  289.76    0.00  3e+38     0.47  8924.05    0.00  3e+38
     8388608       2097152   float     sum    25.69  326.59    0.00  3e+38     0.47  17848.10    0.00  3e+38
    16777216       4194304   float     sum    46.95  357.33    0.00  3e+38     0.47  35696.20    0.00  3e+38
    33554432       8388608   float     sum    89.77  373.80    0.00  3e+38     0.47  71384.81    0.00  3e+38
    67108864      16777216   float     sum    175.5  382.41    0.00  3e+38     0.47  144320.14    0.00  3e+38
   134217728      33554432   float     sum    346.8  386.99    0.00  3e+38     0.47  288640.28    0.00  3e+38
   268435456      67108864   float     sum    689.8  389.14    0.00  3e+38     0.47  571139.27    0.00  3e+38
   536870912     134217728   float     sum   1377.9  389.62    0.00  3e+38     0.51  1052688.06    0.00  3e+38
  1073741824     268435456   float     sum   2749.4  390.54    0.00  3e+38     0.47  2260509.10    0.00  0e+00
# Out of bounds values : 55 FAILED
# Avg bus bandwidth    : 0
```

However it's fine with master.
After some binary search, we found this commit to be the fix:
- https://github.com/NVIDIA/nccl-tests/commit/1ae8cdc315d10c3f65764a1915e0fb0f1563d893

Could you help explain what is the change here and why it's causing the issue on NCCL 2.10.3 + CUDA 11.4, but works fine on NCCL 2.9.8 + CUDA 11.3?

**Comment by AddyLaddy** [comment]
Thanks for your report.
I cannot recreate that failure on my systems with NCCL 2.10.3 + CUDA 11.4. But you said "However it's fine with master."
I am not sure what "master" you are referring too and are you saying it is already resolved?

**Comment by xkszltl** [comment]
> I am not sure what "master" you are referring too and are you saying it is already resolved?

master when filing this issue.

We can stably repro.
Here're docker command you can try:
- Good: `sudo docker run --rm -it --gpus all roasterproject/ubuntu:20210715-f557e4806eb5 bash -c 'set -ex; git clone https://github.com/NVIDIA/nccl-tests; cd nccl-tests; git checkout 1ae8cdc; export LD_LIBRARY_PATH="/usr/local/openmpi/lib:$LD_LIBRARY_PATH"; make -j$(nproc) MPI=0 MPI_HOME="/usr/local/openmpi" NVCC_GENCODE="$(echo 52 60 61 70 75 80 86 | xargs -rn1 | xargs -rI{} echo -gencode=arch=compute_{},code=sm_{} | paste -sd" " -) -gencode=arch=compute_80,code=compute_80"; NCCL_DEBUG=INFO NCCL_NSOCKS_PERTHREAD=8 NCCL_SOCKET_NTHREADS=4 nvidia-smi -L | wc -l | xargs seq | xargs -n1 build/all_reduce_perf -b8 -e1G -f2 -g'`
- Bad: `sudo docker run --rm -it --gpus all roasterproject/ubuntu:20210715-f557e4806eb5 bash -c 'set -ex; git clone https://github.com/NVIDIA/nccl-tests; cd nccl-tests; git checkout 44df0bf; export LD_LIBRARY_PATH="/usr/local/openmpi/lib:$LD_LIBRARY_PATH"; make -j$(nproc) MPI=0 MPI_HOME="/usr/local/openmpi" NVCC_GENCODE="$(echo 52 60 61 70 75 80 86 | xargs -rn1 | xargs -rI{} echo -gencode=arch=compute_{},code=sm_{} | paste -sd" " -) -gencode=arch=compute_80,code=compute_80"; NCCL_DEBUG=INFO NCCL_NSOCKS_PERTHREAD=8 NCCL_SOCKET_NTHREADS=4 nvidia-smi -L | wc -l | xargs seq | xargs -n1 build/all_reduce_perf -b8 -e1G -f2 -g'`

The only diff is the checked out version.
<img width="1115" alt="image" src="https://user-images.githubusercontent.com/5203025/126250783-037e2a68-0c37-49ea-8651-b99c3518e564.png">

Host driver is 450.80.02 with 8 V100, in case that matters.

**Comment by AddyLaddy** [comment]
It looks like I merged in some broken data validation code back in cbe7f65 and it was fixed when I merged in https://github.com/NVIDIA/nccl-tests/commit/1ae8cdc315d10c3f65764a1915e0fb0f1563d893

But I don't understand why this hasn't caused issues before, as that commit was back in April 2019 ?

**Comment by AddyLaddy** [comment]
Maybe I merged in that broken code (locally) a couple of years ago but it got fixed before I pushed the latest update to github?
Is there any reason you are not running the latest master code? The problem does not exist there as you have discovered.

**Comment by xkszltl** [comment]
> Is there any reason you are not running the latest master code?

Because we were using a submodule for nccl-tests and it was created early this year pointing the master at that time.
The latest release tag is from too long ago without some fix we need, so we decided to go with master but need to fix to a version.

Rather than bumping up the version blindly the current master, we'd like to get a clear understanding about what happened, whether it's a bug in cuda, nccl, or nccl-tests, to ensure we're not shadowing potential bugs.

**Comment by xkszltl** [comment]
You may also repro with nvidia/cuda image:
```
sudo docker run --rm -it --gpus all nvidia/cuda:11.4.0-devel-ubuntu20.04 bash -c 'set -ex; rm -rf /etc/apt/sources.list.d/cuda.list; apt-get update; DEBIAN_FRONTEND=noninteractive apt-get install -y git; git clone https://github.com/NVIDIA/nccl-tests; cd nccl-tests; git checkout 44df0bf; export LD_LIBRARY_PATH="/usr/local/openmpi/lib:$LD_LIBRARY_PATH"; make -j$(nproc) MPI=0 MPI_HOME="/usr/local/openmpi" NVCC_GENCODE="$(echo 52 60 61 70 75 80 86 | xargs -rn1 | xargs -rI{} echo -gencode=arch=compute_{},code=sm_{} | paste -sd" " -) -gencode=arch=compute_80,code=compute_80"; NCCL_DEBUG=INFO NCCL_NSOCKS_PERTHREAD=8 NCCL_SOCKET_NTHREADS=4 nvidia-smi -L | wc -l | xargs seq | xargs -n1 build/all_reduce_perf -b8 -e1G -f2 -g'
```

(`rm -rf /etc/apt/sources.list.d/cuda.list` because of https://gitlab.com/nvidia/container-images/cuda/-/issues/133)

**Comment by AddyLaddy** [comment]
Sorry, I was mistaken about the broken commit. The git history is hard to follow as I merged the later fixes on to my bfloat16 branch;
```
git log --graph --pretty=oneline --abbrev-commit
* 1f8f541 (HEAD -> master, origin/master, origin/HEAD) Add CUDA graph support only for CUDA 11.3 and later builds
* b9f90d1 Removed MPI_SUPPORT conditional compilation of average flag
* 547e119 Fix issues with MPI_Allreduce and multi-threaded tests
* 11cff17 Updated with new command line arguments
*   f476f4a (tag: v2.10.0) Merge branch 'bfloat16'
|\  
| * 1dfc76e (bfloat16) Added new option to report average iteration time
| * 1ae8cdc Resync with changes in gitilab-master code
| * 9dae3d3 Added new tests: scatter, sendrecv, hypercube
| * e55ad37 Added support for CUDA graph capture/replay (-G)
| * 526eaca Fixed formatting for bfloat16 support
| * cde7e76 Add support for ncclAvg operation
| * e37545e (origin/bfloat16) Add support for new datatype: bfloat16
* |   44df0bf Merge pull request #88 from nzmsv/master
```

When you checkout 44df0bf you also lose all the later commits. In particular, cde7e76  is necessary for supporting NCCL 2.10.3 as the nccl.h now contains an extra `ncclNumOps` and `ncclNumTypes`.
The nccl-test data validation code prior to that fix doesn't work with NCCL version 2.10.x and later, but if you instead compile it against the NCCL 2.9.9 header it will work.
There is no issue with the NCCL 2.10.x library or CUDA environment, it was simply a bug in the nccl-tests validation code that gets exposed when the `ncclNumOps` and `ncclNumTypes` are changed.

**Comment by xkszltl** [comment]
@AddyLaddy Thanks a lot for help us figure that out!
